mitosheet/mitosheet/ai/recon.py

```diff
@@ -0,0 +1,177 @@
+import ast
+from copy import copy
+from typing import Any, Dict, List, Optional, Tuple
+
+import pandas as pd
+from pandas.testing import assert_frame_equal
+from mitosheet.state import DATAFRAME_SOURCE_AI, State
+from mitosheet.step_performers.column_steps.delete_column import delete_column_ids
+
+from mitosheet.types import ColumnHeader, ColumnReconData, DataframeReconData
+
+def is_df_changed(old: pd.DataFrame, new: pd.DataFrame) -> bool:
+    try:
+        assert_frame_equal(old, new, check_names=False)
+        return False
+    except AssertionError:
+        return True
+
+def exec_for_recon(code: str, original_df_map: Dict[str, pd.DataFrame]) -> DataframeReconData:
+    """
+    Given some Python code, and a list of previously defined dataframes, this function:
+    1. Gets all the newly defined dataframes
+    2. Gets all the dataframes that have been modified
+    3. Gets the value of the last line of code, if that is a Python expression (e.g. not defining a variable)
+
+    It does all of this while only execing the code a single time, which helps performance.
+
+    To find newly defined dataframes, we just save the local variables before the exec runs, 
+    and then find any new dataframes in locals after the exec runs.
+
+    To find the modified dataframes, we capture any dataframe names that are referenced in the code
+    as potentially modified and add them as a local. This then allows us to get out the new values 
+    of that dataframe, and compare them to the old one to see if they really changed.
+
+    Finially, getting the value of the last line requires parsing the Python AST, checking the
+    last expressio, then rebuilding it into a string (aka so we can handle multiple lines) - 
+    and then saving it in fake variable that we can then access again through the locals.
+    """
+    if len(code) == 0:
+        return {
+            'created_dataframes': {},
+            'modified_dataframes': {},
+            'last_line_expression_value': None
+        }
+
+    df_map = {df_name: df.copy(deep=True) for df_name, df in original_df_map.items()}
+    locals_before = copy(locals())
+    ast_before = ast.parse(code)
+
+    last_expression = ast_before.body[-1]
+    if not isinstance(last_expression, ast.Expr):
+        has_last_line_expression_value = False
+    else:
+        has_last_line_expression_value = True
+        last_expression_string = ast.unparse([last_expression]) # type: ignore
+        code = code.replace(last_expression_string, f'FAKE_VAR_NAME = {last_expression_string}')
+    
+    potentially_modified_df_names = [
+        df_name for df_name in original_df_map if 
+        df_name in code
+    ]
+
+    for df_name in potentially_modified_df_names:
+        locals()[df_name] = df_map[df_name]
+
+    exec(code, {}, locals())
```

Comment:
One big missing thing is error handling the exec. We should probably return an inplace Mito error with the last line / full traceback

Reply:
Here are some random things that would probably be worth testing: 
- [ ] invalid Python code 
- [ ] importing a package that is not installed 
- [ ] some long-running operation? 
- [ ] deleting dataframes

Reply:
Will deal with importing a package not installed in a later PR!